tasks/base_tasks.py

```diff
@@ -1592,3 +1610,133 @@ def collect_meta_wrappers(test_module=None, test_all=True):
             yield t, params
             if not test_all:
                 break
+
+
+class CreateRepoTable(Task):
+    schema = 'repo'
+    table = 'filerepo'
+
+    def run(self):
+        session = current_session()
+
+        create_schema = '''
+                        CREATE SCHEMA IF NOT EXISTS "{schema}"
+                        '''.format(schema=self.output().schema)
+        session.execute(create_schema)
+
+        create_table = '''
+                        CREATE TABLE IF NOT EXISTS "{schema}".{table} (
+                            id TEXT,
+                            version NUMERIC,
+                            checksum TEXT,
+                            url TEXT,
+                            path TEXT,
+                            added TIMESTAMP,
+                            PRIMARY KEY(id, version)
+                        );
+                        '''.format(schema=self.output().schema,
+                                   table=self.output().tablename)
+        session.execute(create_table)
+        session.commit()
+
+    def output(self):
+        return PostgresTarget(schema=self.schema, tablename=self.table, non_empty=False)
+
+
+class RemoteDownloader:
+    def download(self, url, output_path):
+        raise NotImplementedError
+
+
+class BaseDownloader(RemoteDownloader):
+    def download(self, url, output_path):
+        urllib.request.urlretrieve(url, output_path)
+
+
+class RepoFile(Task):
+    resource_id = Parameter()
+    url = Parameter()
+    version = IntParameter()
+    downloader = Parameter(default=BaseDownloader())
+
+    _repo_dir = 'repo'
+    _path = None
+
+    def requires(self):
+        return CreateRepoTable()
+
+    def run(self):
+        self.output().makedirs()
+        digest = self._retrieve_remote_file()
+        self._to_db(self.resource_id, self.version, digest, self.url, self.output().path)
+
+    def _create_filepath(self):
+        return self._build_path(str(uuid.uuid4()))
+
+    def _build_path(self, filename):
+        return os.path.join(self._repo_dir, self.resource_id, str(self.version), filename)
+
+    def _retrieve_remote_file(self):
+        if not issubclass(self.downloader.__class__, RemoteDownloader):
+            raise ValueError('downloader must be a subclass of RemoteDownloader')
+
+        self.downloader.download(self.url, self.output().path)
+        return self._digest(self.output().path)
+
+    # https://stackoverflow.com/questions/3431825/generating-an-md5-checksum-of-a-file
+    def _digest(self, file):
+        hash_md5 = hashlib.md5()
+        with open(file, "rb") as f:
+            for chunk in iter(lambda: f.read(4096), b''):
+                hash_md5.update(chunk)
+        return hash_md5.hexdigest()
+
+    def _from_db(self, resource_id, version):
```

Comment:
Have you considered moving these methods related to the table to `CreateRepoTable`? As they only depend on `resource_id` and `version`, `self.input().repo_entry(resource_id, version)` might make sense: it keeps table structure within that file.

Reply:
The Luigi structure is a bit weird: `self.input()` doesn't return `CreateRepoTable` but its output (`PostgresTarget`). We could create a more complex structure to move these functions to `CreateRepoTable` but I think it's not worth it...

Reply:
Oh, I didn't remember that 👍